### PR TITLE
Fix error when pulling one-based data from a zero-based CSR table

### PR DESCRIPTION
**Pull request by Vika-F**

### CODE REVIEW

File: cpp/oneapi/dal/table/backend/csr_kernels.cpp [hunk 1]
```diff
@@ -267,6 +261,7 @@ void pull_csr_block_impl(const Policy& policy,
                                        block_info.indexing_ == sparse_indexing::one_based)
                                           ? 1
                                           : -1);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/backend/csr_kernels.cpp:263]: Zero?

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+#include "oneapi/dal/table/csr.hpp"
+#include "oneapi/dal/table/csr_accessor.hpp"
+#include "oneapi/dal/test/engine/common.hpp"
+
+namespace oneapi::dal {
+
+TEST("can read csr table via csr accessor") {
+    using oneapi::dal::detail::empty_delete;
+
+    double data[] = { 1.0, 2.0, 3.0, 4.0, 1.0, 11.0, 8.0 };
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:26]: What about using templates to test different versions of this functionality?

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                 array<std::int64_t>::wrap(column_indices, element_count),
+                 array<std::int64_t>::wrap(row_offsets, row_count + 1),
+                 column_count };
+    const auto [data_array, cidx_array, ridx_array] = csr_accessor<const double>(t).pull({ 0, -1 });
+
+    REQUIRE(data == data_array.get_data());
+    REQUIRE(column_indices == cidx_array.get_data());
+    REQUIRE(row_offsets == ridx_array.get_data());
+
+    for (std::int64_t i = 0; i < data_array.get_count(); i++) {
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:44]: Minor:
```suggestion
    for (std::int64_t i = 0; i < data_array.get_count(); ++i) {
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        REQUIRE(data_array[i] == data[i]);
+        REQUIRE(cidx_array[i] == column_indices[i]);
+    }
+    for (std::int64_t i = 0; i < ridx_array.get_count(); i++) {
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:48]: Minor:
```suggestion
    for (std::int64_t i = 0; i < ridx_array.get_count(); ++i) {
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+#include "oneapi/dal/test/engine/common.hpp"
+
+namespace oneapi::dal {
+
+TEST("can read csr table via csr accessor") {
+    using oneapi::dal::detail::empty_delete;
+
+    double data[] = { 1.0, 2.0, 3.0, 4.0, 1.0, 11.0, 8.0 };
+    std::int64_t column_indices[] = { 1, 2, 4, 3, 2, 4, 2 };
+    std::int64_t row_offsets[] = { 1, 4, 5, 7, 8 };
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:28]:
```suggestion
    constexpr double data[] = { 1.0, 2.0, 3.0, 4.0, 1.0, 11.0, 8.0 };
    constexpr std::int64_t column_indices[] = { 1, 2, 4, 3, 2, 4, 2 };
    constexpr std::int64_t row_offsets[] = { 1, 4, 5, 7, 8 };
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+    const std::int64_t row_count{ 4 };
+    const std::int64_t column_count{ 4 };
+    const std::int64_t element_count{ 7 };
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:32]:
```suggestion
    constexpr std::int64_t row_count{ 4 };
    constexpr std::int64_t column_count{ 4 };
    constexpr std::int64_t element_count{ 7 };
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        REQUIRE(ridx_array[i] == row_offsets[i]);
+    }
+}
+
+TEST("can read a portion of csr table via csr accessor and create a smaller data block") {
+    using oneapi::dal::detail::empty_delete;
+
+    double data[] = { 1.0, 2.0, 3.0, 4.0, 1.0, 11.0, 8.0 };
+    std::int64_t column_indices[] = { 1, 2, 4, 3, 2, 4, 2 };
+    std::int64_t row_offsets[] = { 1, 4, 5, 7, 8 };
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:58]: Please use `constexpr`

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    const std::int64_t column_count{ 4 };
+    const std::int64_t element_count{ 7 };
+
+    csr_table t{ array<float>::wrap(data, element_count),
+                 array<std::int64_t>::wrap(column_indices, element_count),
+                 array<std::int64_t>::wrap(row_offsets, row_count + 1),
+                 column_count };
+    const auto [data_array, cidx_array, ridx_array] = csr_accessor<const double>(t).pull({ 0, -1 });
+
+    REQUIRE((void*)data != (void*)data_array.get_data());
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:109]: Please use `reinterpret_cast` instead

File: cpp/oneapi/dal/table/test/csr_accessor.cpp (hunk 2 above)
Comment (Vika-F) [reply cpp/oneapi/dal/table/test/csr_accessor.cpp]: Done

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 9]
```diff
@@ -0,0 +1,184 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:2]: Technically it counts as a new file.

**Review by KulikovNikita — APPROVED** [review]

File: cpp/oneapi/dal/table/test/csr_accessor.cpp (hunk 9 above)
Comment (Vika-F) [reply cpp/oneapi/dal/table/test/csr_accessor.cpp]: I've updated the year.

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+                                                                                        8 };
+    static constexpr std::array<std::int64_t, row_count + 1> row_offsets_zero_based_ = { 0,
+                                                                                         3,
+                                                                                         4,
+                                                                                         6,
+                                                                                         7 };
+    const std::int64_t* row_offsets;
+};
+
+using csr_accessor_types = COMBINE_TYPES((int, std::int64_t, float, double), (int, float, double));
```
[the remark this answers is not in the document]
Comment (Vika-F) [reply cpp/oneapi/dal/table/test/csr_accessor.cpp]: Done

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 11]
```diff
@@ -14,100 +14,226 @@
...
+namespace te = dal::test::engine;
+
+/// Tests dal::csr_accessor class on a fixed data table:
+///     | 1,  2,  0,  3 |
+/// A = | 0,  0,  4,  0 |
+///     | 0,  1,  0, 11 |
+///     | 0,  8,  0,  0 |
+/// which is stored in CSR format, zero-based or one-based
+///
+/// @tparam TestType    The puple of two elements.
```
Comment (alexandraepan) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:32]:
```suggestion
/// @tparam TestType    The tuple of two elements.
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 12]
```diff
@@ -14,100 +14,226 @@
...
+    static constexpr std::int64_t row_count = 4; // number of rows in the tested table
+    static constexpr std::int64_t column_count = 4; // number of columns in the tested table
+
+    // number of non-zero elements in the tested table
+    static constexpr std::int64_t element_count = 7;
+
+    /// Check that the test with the particular `TestType` cannot be run on the system
+    /// that does not support 64-bit floating point operations.
+    ///
+    /// @return True, if this test cannot be run on the system.
```
Comment (alexandraepan) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:54]:
```suggestion
    /// @return True if this test cannot be run on the system.
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 13]
```diff
@@ -14,100 +14,226 @@
...
-    REQUIRE(column_indices + 3 == cidx_array.get_data());
-    REQUIRE(row_indices + 1 != ridx_array.get_data());
-
-    for (std::int64_t i = 0; i < data_array.get_count(); i++) {
-        REQUIRE(data_array[i] == data[3 + i]);
-        REQUIRE(cidx_array[i] == column_indices[3 + i]);
-    }
-    for (std::int64_t i = 0; i < ridx_array.get_count(); i++) {
-        REQUIRE(ridx_array[i] == row_indices[1 + i] - 3);
+    /// Initialize column indicies and row offsets for the tested data table in CSR format
```
Comment (alexandraepan) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:62]:
```suggestion
    /// Initialize column indices and row offsets for the tested data table in CSR format
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 14]
```diff
@@ -14,100 +14,226 @@
...
-}
 
-TEST("can read csr table via csr accessor with conversion") {
-    using oneapi::dal::detail::empty_delete;
+    /// Check that `pull` method of `csr_accessor` class works correctly.
+    ///
+    /// @param[in] start_idx    Zero-based index of the first row of the block of data
+    ///                         to be pulled from the table.
+    /// @param[in] end_idx      Either zero-based index of the row that goes after the last row
+    ///                         of the block of data to be pulled from the table,
```
Comment (alexandraepan) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:80]:
```suggestion
    ///                         of the block of data to be pulled from the table;
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 15]
```diff
@@ -14,100 +14,226 @@
...
+    ///                         or -1, if the last row to be pulled is equal to the last row
```
Comment (alexandraepan) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:81]:
```suggestion
    ///                         or -1 if the last row to be pulled is equal to the last row
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 16]
```diff
@@ -14,100 +14,226 @@
...
+    }
 
-    float data[] = { 1.0f, 2.0f, 3.0f, 4.0f, 1.0f, 11.0f, 8.0f };
-    std::int64_t column_indices[] = { 1, 2, 4, 3, 2, 4, 2 };
-    std::int64_t row_indices[] = { 1, 4, 5, 7, 8 };
+    /// Check that the block of rows in CSR format pulled from the table is correct.
+    ///
+    /// @param[in] start_idx    Zero-based index of the first row of the block of data
+    ///                         pulled from the table.
+    /// @param[in] end_idx      Either zero-based index of the row that goes after the last row
```
Comment (alexandraepan) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:106]: same here

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 17]
```diff
@@ -14,100 +14,226 @@
...
+    ///                         of the block of data pulled from the table,
+    ///                         or -1, if the last pulled row is equal to the last row
+    ///                         of the whole table.
+    /// @param[in] data_array   The block of values pulled from the table in the CSR layout.
+    /// @param[in] cidx_array   The block of column indicies pulled from the table
```
Comment (alexandraepan) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:111]:
```suggestion
    /// @param[in] cidx_array   The block of column indices pulled from the table
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 18]
```diff
@@ -14,100 +14,226 @@
...
+            REQUIRE(reinterpret_cast<const table_data_t*>(data) + data_shift ==
+                    reinterpret_cast<const table_data_t*>(data_array.get_data()));
+
+        // check that the data values in the pulled data block are correct
+        for (std::int64_t i = 0; i < data_array.get_count(); i++) {
+            REQUIRE(data_array[i] == data[data_shift + i]);
+        }
+
+        // check column indices and row offsets depending on the indexing schemes
+        // ot the tested table and the accessor
```
Comment (alexandraepan) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:142]:
```suggestion
        // or the tested table and the accessor
```

File: cpp/oneapi/dal/table/test/csr_accessor.cpp [hunk 19]
```diff
@@ -14,100 +14,226 @@
...
+                                                                                         4,
+                                                                                         6,
+                                                                                         7 };
+    const std::int64_t* row_offsets_;
+};
+
+// Generate a sequense of `TestType` template parameters for the `csr_accessor_test`.
+// The first elements in the tuple will be taken from the first list of types;
+// the second elements in the tuple will be taken from the second list of type.
+// As the results, 12 variants of `TestType` parameter are created.
```
Comment (alexandraepan) [inline cpp/oneapi/dal/table/test/csr_accessor.cpp:211]:
```suggestion
// As the result, 12 variants of the `TestType` parameter are created.
```

File: docs/source/conf.py [hunk 20]
```diff
@@ -111,7 +111,7 @@
     'extra_footer': '<p align="right"><a href="https://www.intel.com/content/www/us/en/privacy/intel-cookie-notice.html">Cookies</a></p>'
 }
 
-html_theme_options = { 
+html_theme_options = {
```
Comment (alexandraepan) [inline docs/source/conf.py:114]: This is also addressed in the https://github.com/oneapi-src/oneDAL/pull/2428, so maybe you can skip it

File: cpp/oneapi/dal/table/test/csr_accessor.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Vika-F) [reply cpp/oneapi/dal/table/test/csr_accessor.cpp]: Done

File: cpp/oneapi/dal/table/test/csr_accessor.cpp (hunk 16 above)
Comment (Vika-F) [reply cpp/oneapi/dal/table/test/csr_accessor.cpp:106]: Fixed

File: cpp/oneapi/dal/table/test/csr_accessor.cpp (hunk 18 above)
Comment (Vika-F) [reply cpp/oneapi/dal/table/test/csr_accessor.cpp]: Fixed

**Pull request merged**